hypertrace-trace-enricher/enriched-span-constants/src/main/java/org/hypertrace/traceenricher/enrichedspan/constants/utils/EnrichedSpanUtils.java

```diff
@@ -319,8 +319,8 @@ public static String getUserAgent(Event event) {
     if (event.getHttp() != null && event.getHttp().getRequest() != null) {
       return Optional.ofNullable(event.getHttp().getRequest().getUrl());
     }
-
-    return Optional.empty();
+    String httpUrl = getStringAttribute(event, EnrichedSpanConstants.getValue(Http.HTTP_URL));
```

Comment:
Shouldn't we get rid of reading everything off from first class fields? If the intention is to make this more of a safer change as a fallback, we should be doing the same thing here 

https://github.com/hypertrace/hypertrace-ingester/blob/main/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/HttpFieldsGenerator.java#L427

But, I feel if the intention is to get rid of the first class fields, then let's move towards it, and completely get rid of reading the values from first class fields

Reply:
There is already ongoing effort to clean up all the fields, if i have completely get rid of first class completely then all the logic HttpFieldsGenerator should be copied to here as well, which i don't want to do it. In case if that logic not able to find http url, we can fallback to checking the attribute map which i am doing here.

Reply:
Removal of first class fields is tracked under this issue https://github.com/hypertrace/hypertrace/issues/245, idea is to move logic for populating first class field to a common lib, consumers can then either directly read attributes or leverage this lib

Reply:
> if i have completely get rid of first class completely then all the logic HttpFieldsGenerator should be copied to here as well

Basically, I have couple of concerns

- It's not enough to just read the URLs from `http.url` attribute. It can also be present in `http.request.url` attribute. That's why there is an entire list of attributes here https://github.com/hypertrace/hypertrace-ingester/blob/main/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/HttpFieldsGenerator.java#L78
- We should also validate that it's a full URL, and not a half baked, malformed URL, which is the check here https://github.com/hypertrace/hypertrace-ingester/blob/main/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/HttpFieldsGenerator.java#L432. We should not blindly assume that the attributes will have the URL in the proper format